prefs: add log date format to preferences for git-dag

DasaniT opened the pull request:
I've added the log date format pref discussed in #1312.

davvid commented on cola/widgets/prefs.py:
```diff
@@ -140,6 +142,12 @@ def __init__(self, context, model, parent, source):
+        self.logdate.setToolTip(tooltip)
+        regex = QtCore.QRegExp(logdate.create_regex())
+        validator = QtGui.QRegExpValidator(regex, self.logdate)
+        self.logdate.setValidator(validator)
```
If the setup for the combobox gets too long (it's already longer than the rest) then that should probably be put into a separate `def build_date_formats_combo(parent)` helper function.

davvid commented on docs/git-cola.rst:
```diff
@@ -762,6 +762,12 @@ Defaults to `true`.  This setting is configured using the `Preferences`
 editor's options sub-menu.
 
+cola.logdate
+--------------
```
Make sure that the `---------` line is the exact same length as `cola.logdate` otherwise sphinx will complain about it.

davvid commented on docs/git-cola.rst:
```diff
@@ -762,6 +762,12 @@ Defaults to `true`.  This setting is configured using the `Preferences`
+Set the default date-time mode for the DAG display. Setting a value for
+cola.logdate is similar to using git log's --date option.
+See `git log(1) <https://www.git-scm.com/docs/git-log>`_ for details.
+
```
We should use this URL (with the anchor value) so that we link directly to the `--date=<format>`  docs.

`https://git-scm.com/docs/git-log#Documentation/git-log.txt---dateltformatgt`

https://git-scm.com/docs/git-log#Documentation/git-log.txt---dateltformatgt


Also, instead of "Similar", maybe it should say, "This value is passed to `git log --date=<format>`"?

davvid reviewed (changes requested):
Nicely done. Please squash these sugs into your original commit and force-push over your review branch.

davvid commented:
Kindly adjust the commit to use your full name. There's a short note in `CONTRIBUTING.md` about this:

> Also notice that a real name is used in the Signed-off-by: line. Please don't hide your real name.

DasaniT commented:
Thanks for your feedback! I've made the requested changes, and I made some other small improvements as well.

About the combo, actually at first I did use `qtutils.combo()` and then, as you mentioned, I changed it to text box just because of the `format` option. Right now, the combo box is not editable, because I think that's not necessary. As a result, I removed the validator as well.

DasaniT commented on cola/widgets/prefs.py:
```diff
@@ -140,6 +142,12 @@ def __init__(self, context, model, parent, source):
+        self.logdate.setToolTip(tooltip)
+        regex = QtCore.QRegExp(logdate.create_regex())
+        validator = QtGui.QRegExpValidator(regex, self.logdate)
+        self.logdate.setValidator(validator)
```
About the combo, actually at first I did use `qtutils.combo()` and then, as you mentioned, I changed it to text box just because of the `format:...` option. After the changes, the combo box is not editable, because I think that's not necessary. As a result, I removed the validator as well.

DasaniT commented on cola/logdate.py:
```diff
@@ -0,0 +1,35 @@
+        Option.HUMAN,
+        Option.UNIX,
+        Option.FORMAT,
+    ]
```
I changed "iso" and "iso-strict" to "iso8601" and "iso8601-strict" too.

davvid commented on cola/models/dag.py:
```diff
@@ -273,7 +273,12 @@ def get(self):
+            self._cmd
+            + ['-%d' % self.params.count]
+            + ref_args
+            + ['--date=%s' % self.context.cfg.get('cola.logdate')]
```
sorry for not spotting this earlier (I saw it but forgot to note it).

When we call `get()` please also specify `default=prefs.DateFormat.DEFAULT` and then move the entirety of the `logdate.py` module into `cola/models/prefs.py`. That way we don't have to add a new module just for this small bit of config.

davvid replied:
If we don't specify `default=...` then it'll error out when the user doesn't have `cola.logdate` configured.

DasaniT replied:
I've added a function `logdate(context)' in `cola/models/prefs.py` and I'm calling the function instead of `self.context.cfg.get('cola.logdate'). Is that ok?

And one more thing, I made the change but I would prefer having the configs for logdate in separate file. Because  having them in `cola/models/prefs.py` would make this file messy. There are similar config files like `cola/hidpi.py`.

davvid commented on docs/git-cola.rst:
```diff
@@ -762,6 +762,13 @@ Defaults to `true`.  This setting is configured using the `Preferences`
+------------
+Set the default date-time mode for the DAG display. This value is
+passed to `git log --date=<format>`. See `git log(1) <https://git-scm
+.com/docs/git-log#Documentation/git-log.txt---dateltformatgt>`_ for details.
```
I'd suggest moving the "See ..." line to another line so that we don't have to split the URL across two lines. I'm not sure it'll do the right thing with the newline before the `.com`.

davvid commented:
Very nicely done, thanks!
